master/buildbot/www/hooks/github.py

```diff
@@ -233,6 +346,143 @@ def _process_change(self, payload, user, repo, repo_url, project, event,
 
         return changes
 
+    @defer.inlineCallbacks
+    def _get_http_client(self):
+        if self._http is None:
+            # Instanticate an http client service to user
+            self._http = yield httpclientservice.HTTPClientService.getService(
+                self.master, self.github_api_endpoint, headers={
+                    'Authorization': 'token ' + self.oauth_token,
+                    'User-Agent': 'Buildbot',
+                    'Accept': 'application/vnd.github.v3+json'
+                },
+                debug=self.debug, verify=self.verify)
+            # Get rate limits
+            response = yield self._http.get('/rate_limit')
+            rate_limit_info = yield response.json()
+            core_limits = rate_limit_info['resources']['core']
+            self._api_rate_limit = core_limits['limit']
+            self._api_rate_limit_remaining = core_limits['remaining']
+            self._api_rate_limit_reset = datetime.datetime.utcfromtimestamp(core_limits['reset'])
+            self._log_rate_limit_information()
+        defer.returnValue(self._http)
+
+    def _update_rate_limit_information(self, response):
+        log.info('Response Headers:\n{headers}', headers=pprint.pformat(response.headers))
+        api_rate_limit = response.headers.get('X-RateLimit-Limit') or None
+        if api_rate_limit:
+            self._api_rate_limit = int(api_rate_limit)
+        api_rate_limit_remaining = response.headers.get('X-RateLimit-Remaining') or None
+        if api_rate_limit_remaining:
+            self._api_rate_limit_remaining = int(api_rate_limit_remaining)
+        api_rate_limit_reset = response.headers.get('X-RateLimit-Reset') or None
+        if api_rate_limit_reset:
+            self._api_rate_limit_reset = datetime.datetime.utcfromtimestamp(int(api_rate_limit_reset))
+
+    def _log_rate_limit_information(self):
+        if not self._api_rate_limit or \
+                not self._api_rate_limit_reset or \
+                not self._api_rate_limit_reset:
+            return
+        log.info(
+            'GitHub API rate limit information. Limit: {limit}; '
+            'Remaining: {remaining}; Reset: {reset} UTC;',
+            limit=self._api_rate_limit,
+            remaining=self._api_rate_limit_remaining,
+            reset=self._api_rate_limit_reset)
+
+    @defer.inlineCallbacks
+    def _fetch_pr_data(self, login,
+                       repo_owner=None,
+                       repo_name=None,
+                       pr_num=None):
+
+        httpclient = yield self._get_http_client()
+
+        end_cursor = None
+        ret_change_author = None
+        comments = []
+
+        while True:
```

Comment:
Looks like we need to abstract this query retriever into its own GitHub service. It's too complicated to stay here